bootstrap/kubeadm/api/v1alpha4/kubeadmbootstrapconfig_types.go

```diff
@@ -95,6 +98,31 @@ type KubeadmConfigSpec struct {
 	// For more information, refer to https://github.com/kubernetes-sigs/cluster-api/pull/2763#discussion_r397306055.
 	// +optional
 	UseExperimentalRetryJoin bool `json:"useExperimentalRetryJoin,omitempty"`
+
+	// IgnitionConfig contains Ignition specific configuration.
+	//
+	// If set, bootstrap data will be generated in Ignition format instead of cloud-init.
```

Comment:
If `format` is set to Ignition, default Ignition format will be used for rendering. If one specifies `IgnitionConfig` block instead, then setting `format` becomes optional.

Just one indicator that Ignition is desired is sufficient I think. 

The idea is that `bootstrapDataFormat` function will decide on final format which will be used. So for example, if we add another Ignition flavor in the future, the default may change etc.

Reply:
> then setting format becomes optional.

We shouldn't make it optional, rather always validate that format and IgnitionConfig go hand in hand. If IgnitionConfig is present, we require format = `ignition` and vice-versa.

Reply:
So explicit approach rather than "smart" :+1:

Reply:
Ignition config can be encoded as either `ContainerLinuxConfig` (Ignition config spec v2.x) or `CoreOSConfig` (Ign config spec v3.x).

IMO it makes sense to already add the `CoreOS` config field in the API here now, so the CoreOS config transpilation can be added later without changing it. And make both fields optional.

Reply:
@LorbusChris you sure field for `Ign config spec v3.x` should be called `CoreOSConfig` ? I thought CoreOS is dead. I'd rather add something FCOS specific I guess?

Reply:
https://github.com/coreos/fcct is the transpiler - and there are the `Fedora CoreOS Config` as well as `Red Hat CoreOS Config` flavours of the spec - summing @bgilbert for some bikeshedding on an agnostic name.

Reply:
oh and it's ContainerLinux that's dead, not CoreOS :)

Reply:
As @lorbuschris eludes to above CoreOS lives on with Fedora CoreOS and RHEL CoreOS. If there is a divergence in the usage something like `CoreOSConfig` for spec 3.x and `$GroupUsingItConfig` for 2.x. If we want to leave names out it could simply be `v3Config` and `v2Config` or something like that.

Reply:
I don't have a good name to suggest at the moment. :disappointed:  The FCC config format has `variant` and `version` fields that support selectable front-ends for different Linux distros, and the code for those front-ends can be either in the FCCT codebase or in a different codebase that vendors FCCT.  So it seems awkward to use the name `CoreOS`, since there may eventually be a frontend for a non-"CoreOS" distro.   We might rename FCCT for the same reason; see https://github.com/coreos/fcct/issues/167.

Reply:
It's also important to note that both of these fields would contain sugary configurations in YAML, which first have to be transpiled with their respective transpiler. They're not valid Ignition config JSON in itself.

Reply:
Ah, thanks for that extra context @lorbuschris!

Reply:
I used the `IgnitionConfig` name for the top of this tree, as it contains configuration for generated bootstrap configuration in Ignition format. The property of this Ignition is, that it can be enhanced with CLC/FCOC snippets.

>IMO it makes sense to already add the CoreOS config field in the API here now, so the CoreOS config transpilation can be added later without changing it. And make both fields optional.

Since naming in this case is hard, maybe we can skip it for now? :see_no_evil: Adding it later shouldn't require new API version I guess, as it's not a breaking change or anything.

Reply:
Do we have a concrete suggestion (different from what is here currently) to go forward? It seems like we are unhappy with the naming for future use-cases but there is no consensus for a better name.

Reply:
we use Ignition in house and we don't use any of those formats (clc, or coreOs fcct config), we just just plain yaml and late convert it to the json (which is expected by the Ignition). so effectively what we defined here and what will be genreated and used is a "front end format" (lack of better name).

So consider for future user-cases and customisation, I would suggest something like:

```
type IgnitionFrontendFormat string

const {
 CLC  IgnitionFrontendFormat          = "clc"
 CoreOS  IgnitionFrontendFormat     = "coreOs"
 Customize  IgnitionFrontendFormat = "customize"
}
```

Reply:
@binchenX Would something like `ignition.append` of type `string` work for you, where you could just append your additional Ignition configuration?

Reply:
Note that `append` is terminology specific to Ignition spec 2.  Spec 3 has similar functionality called `merge`, with different semantics.

Reply:
Hmm, I think existing structure as:
```go
// IgnitionSpec contains Ignition specific configuration.
type IgnitionSpec struct {
  // ContainerLinuxConfig contains CLC specific configuration.
  //
  // +optional
  ContainerLinuxConfig *ContainerLinuxConfig `json:"containerLinuxConfig,omitempty"`
}
```

Can be extended further into:
```go
// IgnitionSpec contains Ignition specific configuration.
type IgnitionSpec struct {
  // Frontend defines what configuration will be used to create and customize
  // generated Ignition data.
  Frontend string `json:"frontend,omitempty"`

  // ContainerLinuxConfig contains CLC specific configuration.
  //
  // +optional
  ContainerLinuxConfig *ContainerLinuxConfig `json:"containerLinuxConfig,omitempty"`

  // IgnitionConfig contains direct Ignition configuration, which will be merged or appended to Ignition
  // configuration generated by the bootstrap provider, depending on the specified Ignition version.
  IgnitionConfig *IgnitionConfig
}

type IgnitionConfig struct {
  Version string
  Append string
  Merge string
}
```

I'm not sure if this should be part of the original PR though.

Reply:
I agree this approach is extensible. I don't think it would hurt to include the `IgnitionFrontend` (or similar) field from the start so the next person adding support for another transpiler doesn't have to think about that too much.

Wrt the IgnitionConfig type: It could arguably be just:
```go
type IgnitionConfig struct {
  Version string
  Content string
}
```
or since the version field must be set within the ignition config itself, even this might suffice:
```go
type IgnitionConfig struct {
  Content string
}
```
(I'm not saying that should be part of this PR)

Reply:
For the record, we've now renamed FCCT to [Butane](https://coreos.github.io/butane/), and its input format to [Butane configs](https://coreos.github.io/butane/specs/).

Reply:
@JoelSpeed mind if I mark it as resolved?